### PR TITLE
project refactoring

### PR DESCRIPTION
**Pull request by vanyakosmos**
1. Whole source code was refactored with `black` and linted with `flake8` (its config in `setup.cfg` file).

2. [pre-commit](https://pre-commit.com/) config was added. `pre-commit` allows to run black and flake8 before committing.
  - run `pre-commit run --all-files` to validate all files with black and flake8
  - run `pre-commit install` to create git hook which will run black and flake8 before every commit

3. github actions with linting and coverage report were added

4. tests were slightly refactored. Added new fixture which simplifies test files generation:
```python
docdir.makepyfile("""\
  >>> a = "foo    "
  >>> print(a)
  foo
""")
```
instead of 
```python
p = testdir.makepyfile(
    """
    def foo():
        '''
        >>> a = "foo    "
        >>> print(a)
        foo
        '''
        pass
"""
)
```

and testing:
```python
docdir.test(p, "-W error", passed=1, failed=1)
```
instead of 
```python
out = self.testdir.inline_run(p, "-W error", "--doctest-plus")
out.assertoutcome(passed=1, failed=1)
```

Name of the fixture `docdir` is arguable, but I wasn't able come up with anything else.

## TODOs
- [ ] rename `docdir` into the something sane
- [x] revert black
   - [x] remove string normalization from black config
   - [x] use `"` for human-readable strings and `'` for everything else
   - [x] refactor blocks in parentheses
- [ ] add [coveralls](https://coveralls.io/) integration?

### CODE REVIEW

**Comment by bsipocz** [comment]
I have the firm stand that we shouldn't use black formatting in a library or such. It's OK for user code one needs to review, but here it's makes since horribly looking while also loosing information from the original indenting.

**Comment by vanyakosmos** [comment]
I think that every opensource project should enforce code style, otherwise codebase will eventually become unreadable mess with whole zoo of different styles.

And black being uncompromising formatter is pretty good at enforcing style. It still can be configured tho. For example we can skip string literals normalizations and use already established convention: `"` for human-readable strings and `'` for everything else.

> args being put into new lines with illogical indents, parens in new lines on their own,

Are you talking about this (exactly 1 line in the parenthesis): 
```python
want = re.sub(
  r"(?m)^{}\s*?$".format(re.escape(doctest.BLANKLINE_MARKER)), "", want
)
```
I agree, with only 1 line in `()` it looks kinda ugly. But it can be refactored into this:
```python
pattern = r"(?m)^{}\s*?$".format(re.escape(doctest.BLANKLINE_MARKER))
want = re.sub(pattern, '', want)
```
and in the same way can be refactored another similar code blocks.

> etc

Everything else seems fine to me.

**Comment by astrofrog** [comment]
Thanks for the PR! As @bsipocz mentioned we probably don’t want to use black for now though - I agree consistency is important but for now the most important is consistency across all the repos in the astropy organisation. If we decide to use black or another style convention it would need to be done for all repositories, and it’s a decision we would need to take as a community. So for now let’s keep this open and we can discuss style enforcing libraries at our annual developer meeting in December. If there are other changes you would like to do not related to black maybe you could open a separate PR with those?

**Comment by vanyakosmos** [comment]
what about this:
- i will revert commit with black
- remove black from pre-commit, but keep flake8
- format code with black, but w/o touching string literals
- fix ugly triple liners mentioned by @bsipocz 

anything else?

File: pytest_doctestplus/plugin.py [hunk 1]
```diff
@@ -54,67 +43,110 @@ def __exit__(self, *args, **kwargs):
...
-                        "Options accepted are 'txt', 'tex', and 'rst'. "
-                        "This is no longer recommended, use --doctest-glob instead."
-                     ))
+    parser.addoption(
+        '--doctest-plus',
+        action='store_true',
+        help=(
+            "enable running doctests with additional features not found "
+            "in the normal doctest plugin"
+        ),
```
Comment (bsipocz) [inline pytest_doctestplus/plugin.py]: these are just the black reformattings I was complaining about. The whole section looses the information that it's belong to `parser.addoption` by doing a custom indent, while adds nothing to readability, etc.

Comment (vanyakosmos) [reply pytest_doctestplus/plugin.py]: oh, it wasn't black's fault. I did this by hands 😅 (at least this concrete example)

**Comment by pllim** [comment]
This now has a lot of conflicts. so I am closing this PR.

I am just not sure if we have enough bandwidth to handle massive refactoring PR. Sorry about that.

**Pull request closed without merging**